keps/sig-storage/2324-volume-snapshot-namespace-transfer/README.md

```diff
@@ -0,0 +1,543 @@
+- The VolumeSnapshot must be `readyToUse` before the transfer begins.
+- Any owners of the destination namespace must accept the transfer. By using the StorageTransfer API, this ensures an owner in the destination namespace is aware of the transfer.
+
+## Design Details
+
+This continues the discussion from [API for PVC Namespace Transfer](https://github.com/kubernetes/enhancements/pull/2110/) 
+in regards to the StorageTransfer API.
+
+A `StorageTransferRequest` resource will be created in the source namespace:
+
+```yaml
+apiVersion: storage.k8s.io/v1alpha1
+kind: StorageTransferRequest
+metadata:
+  name: request-name
+  namespace: source-namespace
+spec:
+  source:
+    name: foo
+    kind: VolumeSnapshot
+  secrets:
+    - name: secret-name
+      namespace: secret-namespace
+      type: secret-type # such as `csi.storage.k8s.io/provisioner-secret-name`
+  acceptName: accept-name
+  token: xyji6OwynW # This will be populated automatically by the controller
+  targetName: bar
+```
+
+`StorageTransferAccept` resource will be created in the target namespace:
+
+```yaml
+apiVersion: storage.k8s.io/v1alpha1
+kind: StorageTransferAccept
+metadata:
+  name: accept-name
+  namespace: target-namespace
+spec:
+  sourceNamespace: source-namespace
+  requestToken: xyji6OwynW
+  requestName: request-name
+```
+
+ When matching `StorageTransferRequest` and `StorageTransferAccept` resources are detected, the 
+ transfer process is initiated. We have the following matching criteria for the `StorageTransferRequest` 
+ named `r` and the `StorageTransferAccept` `a`:
+ - `r.metadata.name == a.spec.requestName`
+ - `r.metadata.namespace == a.spec.sourceNamespace`
+ - `r.spec.acceptName == a.metadata.name`
+ - `r.spec.token == a.spec.requestToken`
+
+Once matching is complete, a controller will begin the transfer process. For PVCs, this performs the following:
+- Adds a Finalizer to the PVC source-namespace\foo
+- Set the reclaim policy of the associated PersistentVolume Retain (if not already).
+- Delete the PVC source-namespace\foo.
+- Wait for all Pods to stop using the PVC source-namespace\foo
+- Bind The PersistentVolume to target-namespace\bar.
+- Create the PVC target-namespace\bar by copying the spec of source-namespace\foo and setting spec.volumeName appropriately.
+- Reset the PersistentVolume reclaim policy
+- Remove the Finalizer from the PVC source-namespace\foo
+
+For VolumeSnapshots, this performs the following:
+- Creates a VolumeSnapshot in the target namespace.
+- Binds the VolumeSnapshotContent to the newly created VolumeSnapshot. If this was a dynamically provisioned VolumeSnapshotContent, then it adjusts the VolumeSnapshotContent.Spec.Source to point to the newly created
+VolumeSnapshot.
+- Deletes the source VolumeSnapshot.
+
+The external-snapshotter will need to be updated to remove immutability of the following fields:
```

Comment:
I don't think we should remove the immutability of these fields.  Instead a new VolumeSnapshotContent object should be created.

CC @yuxiangqian